psydac/ddm/cart.py

```diff
@@ -61,7 +61,7 @@ class CartDecomposition():
         (optional: default is MPI_COMM_WORLD).
 
     """
-    def __init__( self, npts, pads, periods, reorder, comm=MPI.COMM_WORLD, nprocs=None, reverse_axis=None ):
+    def __init__( self, npts, pads, periods, reorder, comm=MPI.COMM_WORLD, multiplicity=None, nprocs=None, reverse_axis=None ):
```

Comment:
I am really glad that you found a way to make Psydac work in parallel with knot sequences of non-unit multiplicity, but I think that it is excessive to have the `multiplicity` integer values appear also in this module, whose only purpose is to distribute an N-dimensional `StencilVector` across multiple processes. Do we really need to specialize `CartDecomposition` so much?

What are we going to do if we decide to support knot sequences where each knot can have a different multiplicity? Are we going to change every class in Psydac once again?

Reply:
if it was just one space, I would not have included this information in the cart, the problem is that we can have different spaces with different multiplicities , so we need to make sure that the decomposition makes sense.
when we use the method `reduce_elements`  to create the Derham sequence , we can reduce the space by just `1` element or `n` elements or possibly `m*n` such that `m` is the multiplicity and `n` the number of breakpoints.
we need to make a decomposition such that the break points owned by  each process unique and are identical between  between the spaces

Reply:
in the case of a general knot sequence, we will for sure need to change everything , maybe not as much as in this case, but we will need to update the generated assembly code, the linear algebra methods like (dot, transpose), maybe the data structure, and also the decomposition need to make sense so that the breakpoints are unique in each process

Reply:
Since we are distributing the B-spline coefficients, and hence the basis functions of a given spline space, we can think of assigning to each process a different knot sequence, which defines a different set of B-splines. By so doing `p + 1` knots are repeated between adjacent processes, and if the multiplicity of every interior knot is `1` there will be `p + 1` repeated breakpoints.

Given the above, I cannot see what you mean by "distributing the breakpoints" and making them "unique in each process". Do you mean that you want the breakpoints on a given process to remain unchanged after reducing the space degree?

@ratnania We may need a short discussion is needed in order to understand each other.

Reply:
yes that's what I mean

Reply:
OK.

When we assemble a mass (or stiffness) matrix by adding "element mass matrices", we are essentially decomposing an integral over the domain into a sum of integrals over each element. The requirement of having domain and codomain share the same breakpoints on a given process seems to simplify this process... is that the reason why you want that property?

Reply:
yes this is one of the reasons, but also for the dot product to work we need that , otherwise, we will need more ghost regions